File: apps/web/lib/api/tools.ts
import { api, queryString, type QueryParams } from "./client";
import type { PaginatedItems, ToolDefinition } from "@/lib/types";

export type ToolRunStatus =
  | "planned" | "awaiting_input" | "awaiting_confirmation" | "queued" | "running"
  | "retry_scheduled" | "succeeded" | "failed" | "cancelled" | "blocked" | "timed_out";

export interface ToolRun {
  id: string;
  user_id: string;
  companion_id: string;
  conversation_id?: string | null;
  tool_definition_id?: string | null;
  parent_tool_run_id?: string | null;
  requested_by?: string | null;
  trace_run_id?: string | null;
  capability?: string | null;
  status: ToolRunStatus;
  risk_level: string;
  permission_required: boolean;
  permission_granted: boolean;
  confirmation_required: boolean;
  confirmation_summary?: string | null;
  input_json: Record<string, unknown>;
  output_json: Record<string, unknown>;
  error_json: Record<string, unknown>;
  evidence_refs: Array<Record<string, unknown>>;
  attempt_count: number;
  max_attempts: number;
  timeout_seconds: number;
  terminal_reason?: string | null;
  request_message_id?: string | null;
  result_message_id?: string | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  elapsed_ms?: number | null;
}

export type ToolPermissionPolicy = "not_required" | "ask_once" | "ask_every_time" | "disabled";

export interface ToolPermission {
  id: string;
  user_id: string;
  companion_id: string;
  tool_definition_id: string;
  policy: ToolPermissionPolicy;
  status: "active" | "denied" | "revoked" | "expired";
  allowed_until?: string | null;
  reason?: string | null;
  scope_json?: Record<string, unknown>;
}

export interface ToolRunActionScope {
  companion_id: string;
  conversation_id?: string | null;
  reason?: string;
}

export interface ToolResource {
  id: string;
  companion_id: string;
  conversation_id?: string | null;
  resource_type: "reminder" | "calendar_event" | "note";
  title: string;
  content?: string | null;
  status: "active" | "completed" | "cancelled" | "archived";
  starts_at?: string | null;
  due_at?: string | null;
  timezone?: string | null;
  resource_json: Record<string, unknown>;
  created_at?: string | null;
}

export function listToolDefinitions(params?: QueryParams) {
  return api.get<PaginatedItems<ToolDefinition>>(`/tool-definitions${queryString(params)}`);
}

export function getToolDefinition(id: string) {
  return api.get<ToolDefinition>(`/tool-definitions/${id}`);
}

export function listToolPermissions(params: QueryParams & { companion_id: string }) {
  return api.get<PaginatedItems<ToolPermission>>(`/tool-permissions${queryString(params)}`);
}

export function updateToolPermission(
  id: string,
  data: { companion_id: string; policy?: ToolPermissionPolicy; status?: ToolPermission["status"]; allowed_until?: string | null; reason?: string; scope_json?: Record<string, unknown> },
) {
  return api.patch<ToolPermission>(`/tool-permissions/${id}`, data);
}

export function setToolPermission(
  toolDefinitionId: string,
  data: { companion_id: string; policy: ToolPermissionPolicy; reason?: string },
) {
  return api.put<ToolPermission>(`/tool-permissions/by-definition/${toolDefinitionId}`, data);
}

export function listToolRuns(params: QueryParams & { companion_id: string }) {
  return api.get<PaginatedItems<ToolRun>>(`/tool-runs${queryString(params)}`);
}

export function getToolRun(id: string, companionId: string) {
  return api.get<ToolRun>(`/tool-runs/${id}${queryString({ companion_id: companionId })}`);
}

export function confirmToolRun(id: string, scope: ToolRunActionScope) {
  return api.post<ToolRun>(`/tool-runs/${id}/confirm`, scope);
}

export function cancelToolRun(id: string, scope: ToolRunActionScope) {
  return api.post<ToolRun>(`/tool-runs/${id}/cancel`, scope);
}

export function retryToolRun(id: string, scope: ToolRunActionScope) {
  return api.post<ToolRun>(`/tool-runs/${id}/retry`, scope);
}

export function createToolRunBadCase(id: string, scope: ToolRunActionScope) {
  return api.post(`/tool-runs/${id}/create-bad-case`, scope);
}

export function listToolResources(params: QueryParams & { companion_id: string }) {
  return api.get<ToolResource[]>(`/tool-resources${queryString(params)}`);
}

File: services/agent-api/app/api/routes/tools.py
"""Bounded Tool Runtime routes."""

import uuid

from fastapi import APIRouter, Query

from app.schemas.common import err, ok, paginated_ok
from app.schemas.tool import ToolPermissionSet, ToolPermissionUpdate, ToolRunActionRequest
from app.services import tool_runtime_service, tool_service
from app.services.conversation_service import ConversationTurnError


router = APIRouter(tags=["Tools"])


@router.get("/tool-definitions")
def list_tool_definitions(page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), companion_id: str | None = None):
    result = tool_service.list_tool_definitions(page, page_size, companion_id=uuid.UUID(companion_id) if companion_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.get("/tool-definitions/{tool_definition_id}")
def get_tool_definition(tool_definition_id: str):
    row = tool_service.get_tool_definition(uuid.UUID(tool_definition_id))
    return ok(row) if row else err("NOT_FOUND", "Tool definition not found")


@router.get("/tool-permissions")
def list_tool_permissions(companion_id: str = Query(...), page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), tool_definition_id: str | None = None):
    result = tool_service.list_tool_permissions(page, page_size, companion_id=uuid.UUID(companion_id), tool_definition_id=uuid.UUID(tool_definition_id) if tool_definition_id else None)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.patch("/tool-permissions/{permission_id}")
def update_tool_permission(permission_id: str, body: ToolPermissionUpdate):
    row = tool_service.update_tool_permission_scoped(uuid.UUID(permission_id), body.companion_id, body.model_dump(exclude={"companion_id"}, exclude_none=True))
    return ok(row) if row else err("NOT_FOUND", "Tool permission not found in this Companion scope")


@router.put("/tool-permissions/by-definition/{tool_definition_id}")
def set_tool_permission(tool_definition_id: str, body: ToolPermissionSet):
    row = tool_service.set_tool_permission_scoped(
        uuid.UUID(tool_definition_id),
        body.companion_id,
        body.policy,
        body.reason,
    )
    return ok(row) if row else err("NOT_FOUND", "Tool definition or Companion not found in this scope")


@router.get("/tool-runs")
def list_tool_runs(companion_id: str = Query(...), conversation_id: str | None = None, page: int = Query(1, ge=1), page_size: int = Query(20, ge=1, le=100), status: str | None = None):
    result = tool_service.list_tool_runs(page, page_size, companion_id=uuid.UUID(companion_id), conversation_id=uuid.UUID(conversation_id) if conversation_id else None, status=status)
    return paginated_ok(result["items"], page, page_size, result["total"])


@router.get("/tool-runs/{tool_run_id}")
def get_tool_run(tool_run_id: str, companion_id: str = Query(...)):
    row = tool_service.get_tool_run(uuid.UUID(tool_run_id), uuid.UUID(companion_id))
    return ok(row) if row else err("NOT_FOUND", "ToolRun not found in this Companion scope")


@router.post("/tool-runs/{tool_run_id}/confirm")
def confirm_tool_run(tool_run_id: str, body: ToolRunActionRequest):
    try:
        row = tool_runtime_service.confirm_tool_run(uuid.UUID(tool_run_id), companion_id=body.companion_id, conversation_id=body.conversation_id)
    except ConversationTurnError as exc:
        return err(exc.code, exc.message, exc.details)
    return ok(row) if row else err("NOT_FOUND", "ToolRun not found")


@router.post("/tool-runs/{tool_run_id}/cancel")
def cancel_tool_run(tool_run_id: str, body: ToolRunActionRequest):
    try:
        row = tool_runtime_service.cancel_tool_run(uuid.UUID(tool_run_id), companion_id=body.companion_id, conversation_id=body.conversation_id, reason=body.reason)
    except ConversationTurnError as exc:
        return err(exc.code, exc.message, exc.details)
    return ok(row) if row else err("NOT_FOUND", "ToolRun not found")


@router.post("/tool-runs/{tool_run_id}/retry")
def retry_tool_run(tool_run_id: str, body: ToolRunActionRequest):
    try:
        row = tool_runtime_service.retry_tool_run(uuid.UUID(tool_run_id), companion_id=body.companion_id, conversation_id=body.conversation_id)
    except ConversationTurnError as exc:
        return err(exc.code, exc.message, exc.details)
    return ok(row) if row else err("NOT_FOUND", "ToolRun not found")


@router.post("/tool-runs/{tool_run_id}/create-bad-case")
def create_tool_run_bad_case(tool_run_id: str, body: ToolRunActionRequest):
    row = tool_service.create_tool_run_bad_case(uuid.UUID(tool_run_id), {"description": body.reason} if body.reason else None, companion_id=body.companion_id)
    return ok(row) if row else err("NOT_FOUND", "ToolRun not found in this Companion scope")


@router.get("/tool-resources")
def list_tool_resources(companion_id: str = Query(...), resource_type: str | None = None, status: str | None = None):
    return ok(tool_runtime_service.list_tool_resources(companion_id=uuid.UUID(companion_id), resource_type=resource_type, status=status))
